### PR TITLE
DAOS-4717 pool: per VOS pool GC ULT

### PR DESCRIPTION
**Pull request by NiuYawei**
Change GC ULTs from per VOS xstream to per VOS pool, so that pool GC
ULTs can be scheduled independently.

Disable container destroy sync wait to avoid RPC timeout on huge
container destroy.

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### CODE REVIEW

File: src/container/srv_target.c [hunk 1]
```diff
@@ -166,6 +172,7 @@ cont_child_aggregate(struct ds_cont_child *cont, uint64_t *msecs)
 
 	if (epoch_min > epoch_max) {
 		/* Nothing can be aggregated */
+		*msecs = max(*msecs, (epoch_min - epoch_max) / 1000000);
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/container/srv_target.c:163]: sure.

File: src/vos/vos_gc.c [hunk 2]
```diff
@@ -878,12 +880,84 @@ vos_gc_pool(daos_handle_t poh, int *credits)
 	}
 	total -= *credits; /* subtract the remained credits */
 
-	if (empty && total != 0) /* did something */
-		gc_log_pool(pool);
+	if (empty) {
+		if (total != 0) /* did something */
+			gc_log_pool(pool);
+		gc_del_pool(pool);
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/vos/vos_gc.c:886]: It'll be re-added on container, object, key record deletion.

File: src/vos/vos_gc.c [hunk 3]
```diff
@@ -878,12 +880,84 @@ vos_gc_pool(daos_handle_t poh, int *credits)
...
+	struct vos_pool	*pool = vos_hdl2pool(poh);
+	int		 rc, total = 0;
+
+	D_ASSERT(!daos_handle_is_inval(poh));
+	if (!gc_have_pool(pool))
+		return 0; /* nothing to reclaim for this pool */
+
+	/*
+	 * Pause flushing free extents in VEA aging buffer, otherwise,
+	 * there'll be way more fragments to be processed.
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/vos/vos_gc.c:915]: The freed extents won't be available for allocation instantly, it'll stay in an aging buffer and being coalesced, once the coalesced free extent expired, it'll be inserted into the in-memory free extent tree and being available for allocation. The purpose is to reduce fragmentation.

Here when we do this GC loop, it's likely that huge amount of free extents could be generated (when some container is destroyed), so we 'plug' the aging buffer temporarily, and 'unplug' it after the loop finished, this plug & unplug can make sure all free extents being merged into few giant free extents (when on container destroy case)

File: src/vos/vos_gc.c [hunk 4]
```diff
@@ -878,12 +880,84 @@ vos_gc_pool(daos_handle_t poh, int *credits)
...
+		total -= creds; /* subtract the remainded credits */
+		if (creds != 0)
+			break; /* reclaimed everything */
+
+		if (credits > 0 && total >= credits)
+			break; /* consumed all credits */
+
+		if (vos_gc_yield(yield_func, yield_arg)) {
+			D_DEBUG(DB_TRACE, "GC pool run aborted\n");
+			rc = 1;
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/vos/vos_gc.c:940]: Not much difference, caller can know if it was aborted by the return value.

**Comment by ashleypittman** [comment]
I've checked the errors logged by run_test and it's a artifact of the fact we fixed some errors in master recently, and this PR is based on a commit before but is using a reference build after so they can be ignored.

File: src/pool/srv_target.c [hunk 5]
```diff
@@ -107,6 +107,84 @@ ds_pool_child_purge(struct pool_tls *tls)
...
+		if (rc < 0)
+			D_ERROR(DF_UUID"[%d]: GC pool run failed. "DF_RC"\n",
+				DP_UUID(child->spc_uuid), dmi->dmi_tgt_id,
+				DP_RC(rc));
+
+		if (dss_ult_exiting(child->spc_gc_req))
+			break;
+
+		/* Sleep 2 seconds */
+		sched_req_sleep(child->spc_gc_req, 2ULL * 1000);
```
Comment (gnailzenh) [inline src/pool/srv_target.c:134]: I think this should be improved in the future, e.g: based on SCM space pressure, I/O workload...

Comment (NiuYawei) [reply src/pool/srv_target.c:134]: Yes, I think ideally the GC and aggregation ULT (especially GC ULT) should always in sleep unless it's woke up by somebody else (on container destroy or aggregation progressed). I'm going to enlarge this sleep period in follow-on patch.

**Pull request merged**